geonotebook/vis/ktile/ktile.py

```diff
@@ -0,0 +1,233 @@
+
+    def _dynamic_vrt_options(self, data, kwargs):
+        options = {
+            'path': os.path.abspath(data.path),
+            'bands': data.band_indexes,
+
+            'nodata': data.nodata,
+            # TODO:  Needs to be moved into RasterData level API
+            'raster_x_size': data.reader.width,
+            'raster_y_size': data.reader.height,
+            'transform': data.reader.dataset.profile['transform'],
+            'dtype': data.reader.dataset.profile['dtype']
+        }
+        if 'map_srs' in kwargs:
+            options['map_srs'] = kwargs['map_srs']
+
+        return options
+
+    def ingest(self, data, name=None, **kwargs):
+
+        # Verify that a kernel_id is present otherwise we can't
+        # post to the server extension to add the layer
+        kernel_id = kwargs.pop('kernel_id', None)
+        if kernel_id is None:
+            raise Exception(
+                "KTile vis server requires kernel_id as kwarg to ingest!")
+
+        options = {
+            'name': data.name if name is None else name
+        }
+
+        options.update(kwargs)
+
+        # Note:
+        # Check if the reader has defined a vrt_path
+        #
+        # This is mostly intended for the VRTReader so that it can communicate
+        # that the VRT for reading data is also the VRT that should be used for
+        # visualisation. Otherwise we wouild have to explicitly add a vrt_path
+        # kwarg to the add_layer() call.
+        #
+        # A /different/ VRT can still be used for visualisation by passing
+        # a path via vrt_path to add_layer.
+        #
+        # Finally, A dynamic VRT will ALWAYS be generated if vrt_path is
+        # explicitly set to None via add_layer.
+        if hasattr(data.reader, 'vrt_path'):
+            if 'vrt_path' in kwargs and kwargs['vrt_path'] is None:
+                # Explicitly set to None
+                pass
+            else:
+                kwargs['vrt_path'] = data.reader.vrt_path
+
+        # If we have a static VRT
+        if 'vrt_path' in kwargs and kwargs['vrt_path'] is not None:
+            options.update(self._static_vrt_options(data, kwargs))
+        else:
+            # We don't have a static VRT, set options for a dynamic VRT
+            options.update(self._dynamic_vrt_options(data, kwargs))
+
+        # Make the Request
+        base_url = '{}/{}/{}'.format(self.base_url, kernel_id, name)
+
+        r = requests.post(base_url, json={
+            "provider": {
+                "class": "geonotebook.vis.ktile.provider:MapnikPythonProvider",
+                "kwargs": options
+            }
+            # NB: Other KTile layer options could go here
+            #     See: http://tilestache.org/doc/#layers
+        })
+
+        try:
+            response = r.json()
+        except Exception:
+            response = {
+                'status': 0,
+                'error': str(r)
+            }
+
+        try:
+            if response['status'] == 0:
```

Comment:
I think this will be [possible](http://stackoverflow.com/questions/13730945/how-do-i-return-http-error-code-without-default-template-in-tornado),  I just need to return JSON from tornado endpoint on error so we can get access to the stacktrace from the tornado process inside the kernel (so it can be forwarded to the browser)